contents/regex/Regex.md

```diff
@@ -0,0 +1,367 @@
+
+## Search and Replace
+
+Regex can also be used to search and replace strings.
+
+Say you have a passage of text which contains NRICs to be censored.
+
+### Example
+
+> * John - S1234567A
+> * Mary - S8192853B
+> * David - S1235326C
+
+This can't be done with a simple search and replace, since NRICs have different digits, and we can't specify which ones to look for. However, we know that an NRIC consists of an S, 7 digits from 0-9 and ends with a letter. Knowing this we can construct a Reges:
+
+* Find: `S[\d]{7}[A-Z]`
+* Replace: `SXXXXXXXX`
+
+As we can see, the `S` matches the very first S, [\d]{7} matches a string of 7 digits, and [A-Z] matches a single capital letter. We then replace all the critical information with a string of Xs.
+
+### Result
+
+> * John - SXXXXXXXX
+> * Mary - SXXXXXXXX
+> * David - SXXXXXXXX
+
+## Capturing groups
+
+Now, imagine you have a section of text with many dates written in the US Date Format (MM-DD-YYYY).
+
+> Calendar
+> * 02-20-2017 Paul's Birthday
+> * 02-24-2017 Presentation at work
+> * 02-25-2017 Workshop
+
+However, you want them to be written in (DD-MM-YYYY) instead. As we can see, this isn't as simple as just finding and replacing text, we must also maintain the data in the search string (e.g. `02-20` becomes `20-02`). This can be done using a **capturing group**. A capturing group is a part of a Regex surrounded by brackets `()`. They are numbered `1`, `2`, `3`, from left to right and are used to hold captured data. For example:
+
+`(\d\d)-(\d\d)-(\d\d\d\d)` against `02/20/2017` will result in group 1 : `02`, group 2 : `20`, and group 3 : `2017`.
+
+Using this, we can construct our regex as follows:
+
+* Find: `(\d\d)-(\d\d)-(\d\d\d\d)`
+* Replace: `\2-\1-\3`
+
+In the replace field, `\1` represents the group matched in the string. For example, `\1` corresponds to `02`, `\2` corresponds to `20` and so on. As you can see, we are swapping group 1 (the month) and group 2 (the day) to get the desired result.
+
+### Result
+
+> Calendar <br>
+> * 20-02-2017 Paul's Birthday <br>
+> * 24-02-2017 Presentation at work <br>
+> * 25-02-2017 Workshop
+
+## Further exploration
+
+These examples provide a quick look as to what Regex can be used for, without code. You can also use Regex as part of many different programming languages, to search strings within the code. Many languages implement it differently, for example Python uses the [`re`](https://docs.python.org/2/library/re.html) library, Java uses [`Pattern`](http://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html) and Javascript has regex built into the language using [`/` notation](https://developer.mozilla.org/en/docs/Web/JavaScript/Guide/Regular_Expressions). You can search the respective language documentation for a more comprehensive view of how to use Regex in the language of your choice.
```

Comment:
`Many languages implement it differently` : this sounds like a negative thing. Rephrase?